### ISSUE TITLE
Figure out USB-TMC issues

### ISSUE THREAD

**Issue by jorisvr** [teamdiamond/qmi#37]
Our methods for communicating with USBTMC devices are difficult and/or unreliable.

So far we have tested the following methods:
1. pyvisa with NI-VISA under Windows.
Basic SCPI commands work.
An exception may occur in case of command timeout, but the system appears to recover and responds correctly to subsequent commands.

2. pyvisa with pyvisa-py under Linux.
(Note: pyvisa-py is a python/libusb visa backend, alternative for NI-VISA).
The Debian stable release of pyvisa-py has a severe bug. Get pyvisa-py from its upstream github.
Basic SCPI commands work.
After a timeout exception, further commands are not processed reliably. This is seen especially with the Tektronix frequency counter, which seems to become fully unresponsive after a timeout.

3. usb-tmc kernel driver under Linux.
Naive attempts to communicate via read/write to /dev/usbtmc0 fail (Joris).
Careful sequencing of I/O operations to /dev/usbtmc0 leads to reliable operation (Sidney).
Operation with the Tektronix seems more reliable with this method than with pyvisa-py. However the device still becomes unresponsive after timeout.

TODO:
* Confirm that NI-VISA recovers reliably from timeout with Tektronix.
* Debug USB communication and find differences between NI-VISA vs pyvisa-py vs kernel driver.
Check USB-TMC spec and figure out what is going on.
* Create a reliable solution based on kernel driver or pyvisa-py or custom Python code.
(Note: Avoid depending on a patched kernel driver.)

**Comment by jorisvr** [comment]
Test results:
* Indeed NI-VISA also raises exception on timeout, but always recovers and handles subsequent commands correctly.
* Both pyvisa-py and usbtmc driver consistently fail after timeout. Power-cycling of device is needed to recover.

USB traces:
* NI-VISA explicitly sends INITIATE_ABORT_BULK_IN (per USBTMC specification) after timeout to clear the pending read operation in the device.
* Pyvisa-py and the kernel driver do not send this command. As a consequence, the read operation remains pending in the device and the device is unable to handle further commands until this is resolved. (Devices are explicitly allowed to refuse further operations while a read command is pending.)

Conclusion:
* NI-VISA behaves correctly per USBTMC spec.
* The Tektronix FCA3000 behaves correctly per USBTMC spec.
* Pyvisa-py is incorrect. It fails to send an abort command to terminate a pending read operation after it has decided to time-out the read.
* Host behaviour with the kernel driver is also incorrect, but this is not necessarily a bug in the driver. The driver can optionally be configured to automatically abort after timeout or the application can explicitly trigger an abort sequence via ioctl. This still needs to be tested.

**Comment by jorisvr** [comment]
Conclusion:
* NI-VISA works ok (under Windows).
* We patched pyvisa-py to add an explicit abort sequence after timeout and confirmed reliable recovery from timeout. Patch is pending upstream: https://github.com/pyvisa/pyvisa-py/pull/179
* python-usbtmc (a standalone USBTMC implementation in Python) contained an incorrect implementation of the abort sequence which did not work. We patched python-usbtmc to fix the abort sequence and confirmed reliable recovery from timeout. Patch is pending upstream: https://github.com/python-ivi/python-usbtmc/pull/50
* The Linux kernel usbtmc driver can probably be made to recover from timeout via the proper ioctls etc. However the driver has a fixed 5 second timeout which can not be configured.

We resolve this issue for now by using python-usbtmc under Linux (with our patch).
Our QMI_UsbTmcTransport automatically maps to python-usbtmc under Linux (and to pyvisa with NI-VISA under Windows).
